scripts/lib/api/conversionPipeline.ts

```diff
@@ -0,0 +1,194 @@
+import { readFile, writeFile } from "fs/promises";
+
+import { globby } from "globby";
+import { uniq, uniqBy } from "lodash";
+import { mkdirp } from "mkdirp";
+import transformLinks from "transform-markdown-links";
+
+import { ObjectsInv } from "./objectsInv";
+import { sphinxHtmlToMarkdown } from "./htmlToMd";
+import { saveImages } from "./saveImages";
+import { generateToc } from "./generateToc";
+import { HtmlToMdResultWithUrl } from "./HtmlToMdResult";
+import { mergeClassMembers } from "./mergeClassMembers";
+import flattenFolders from "./flattenFolders";
+import { updateLinks } from "./updateLinks";
+import { specialCaseResults } from "./specialCaseResults";
+import addFrontMatter from "./addFrontMatter";
+import { dedupeHtmlIdsFromResults } from "./dedupeHtmlIds";
+import { Pkg } from "./Pkg";
+import { pathExists, getRoot } from "../fs";
+import {
+  addNewReleaseNotes,
+  generateReleaseNotesIndex,
+  updateHistoricalTocFiles,
+  writeReleaseNoteForVersion,
+} from "./releaseNotes";
+
+export async function runConversionPipeline(
+  htmlPath: string,
+  markdownPath: string,
+  pkg: Pkg,
+) {
+  const maybeObjectsInv = await (pkg.hasObjectsInv()
+    ? ObjectsInv.fromFile(htmlPath)
+    : undefined);
+  const files = await globby(
+    [
+      "apidocs/**.html",
+      "apidoc/**.html",
+      "stubs/**.html",
+      "release_notes.html",
+    ],
+    {
+      cwd: htmlPath,
+    },
+  );
+
+  let results: Array<HtmlToMdResultWithUrl> = [];
+  for (const file of files) {
+    const html = await readFile(join(htmlPath, file), "utf-8");
+    const result = await sphinxHtmlToMarkdown({
+      html,
+      fileName: file,
+      determineGithubUrl: pkg.determineGithubUrlFn(),
+      imageDestination: pkg.outputDir("/images"),
+      releaseNotesTitle: `${pkg.title} ${pkg.versionWithoutPatch} release notes`,
+    });
+
+    // Avoid creating an empty markdown file for HTML files without content
+    // (e.g. HTML redirects)
+    if (result.markdown == "") {
+      continue;
+    }
+
+    const { dir, name } = parse(`${markdownPath}/${file}`);
+    let url = `/${relative(`${getRoot()}/docs`, dir)}/${name}`;
+    results.push({ ...result, url });
+  }
+
+  const allImages = uniqBy(
+    results.flatMap((result) => result.images),
+    (image) => image.fileName,
+  );
+
+  const dirsNeeded = uniq(
+    results.map((result) => parse(urlToPath(result.url)).dir),
+  );
+  for (const dir of dirsNeeded) {
+    await mkdirp(dir);
+  }
+
+  results = await mergeClassMembers(results);
+  flattenFolders(results);
+  specialCaseResults(results);
+  await updateLinks(results, maybeObjectsInv);
+  await dedupeHtmlIdsFromResults(results);
+  addFrontMatter(results, pkg);
+
+  await maybeObjectsInv?.write(pkg.outputDir("public"));
+  for (const result of results) {
+    let path = urlToPath(result.url);
+
+    if (path.endsWith("release-notes.md")) {
+      const shouldUse = await handleReleaseNotesFile(result, pkg);
+      if (!shouldUse) continue;
```

Comment:
Nit: What about changing the `shouldUse` name for something like `shouldWrite` or `writeSingleFile`? I think it will make it clear that we are checking whether to write a single release note file or not in the `writeFile` call of line 111